### DISCUSSION TITLE
Pack all code in a class

### DISCUSSION THREAD

**Discussion by nguyenpham** [official-stockfish/Stockfish#4797] [Ideas]
Recently all SF’s code is packed in a namespace. IMO, it is a good improvement. I wonder if we could put it further, to pack all code in a class. Say, a class named Stockfish. The main function will look like:
```
int main() {
   Stockfish stockfish;
   stockfish.init();
   stockfish.loop(argc, argv);
}
```
It will take all advantages/benefits as having a namespace, plus some benefits from being able to create and destroy instances of the class:
- easier to debug since all variants, functions are in an instance
- easier for some purposes such as developing library or to sharing code with other engines/apps. For example, an iOS app may have several chess engines. All code of those engines must be embedded directly into the app. Whenever users change engines, the previous one should be removed from the memory and the new chess engine should be created. The easiest way is to create/destroy  instances of their main classes

**Comment by vondele** [comment]
I think there is some overlap with the idea from https://github.com/official-stockfish/Stockfish/pull/4626 ideally you could even create a game manager that plays against two instances just created like in your above snippet.

**Comment by Sopel97** [comment]
Way too much global state for this to be useful. You would still be able to only make 1 instance.

**Reply by vondele** [reply]
right, would require significant work to remove all global state (most notably, TT and histories, but there is more)

**Reply by vondele** [reply]
would be beneficial, for example, to write light-weight datagen code for training, avoiding maintaining a separate tools branch.

**Reply by nguyenpham** [reply]
We may solve that problem by two steps. First, pack all global variants into the SF class as static ones thus all code could be compiled and run quickly. Those variants could be removed/deleted manually by a function if we want. After that we may improve one by one, change from static to normal variants and/or make sure they could be created/removed properly.

**Reply by Disservin** [reply]
> would be beneficial, for example, to write light-weight datagen code for training, avoiding maintaining a separate tools branch.

Mhh, datagen still requires some changes here and there and a few engine devs disable certain pruning heuristics in search, so that light weight datagen code would only be temporarily I think.

**Comment by jshriver** [comment]
Curious what is the current branch for datagen?  I thought all of the
nodechip stuff was deprecated since the merger of lc0 data around SF14 or
15.

**Reply by Disservin** [reply]
It's in the tools branch, https://github.com/official-stockfish/Stockfish/tree/tools.
But it's probably a long time ago since someone has generated some data using Stockfish.

**Comment by jshriver** [comment]
Aye, at one point I generated  just over 100gigs of binpacks but gave up
when I was told they were no longer being used. Instead previous ones,
nets, merged with the lc0 data and I guess since then it's the same net
just constant updates based on previous data.  Was fun at the time and
enjoyed rolling NN's.  The process seems past me now or at least more
closed off.

I still have a copy of those but figure they are somewhat pointless since
they were SF14 maybe early SF15.
